src/ecc.rs

```diff
@@ -437,8 +457,104 @@ impl Mechanism for EccMechanism {
     }
 }
 
+fn kdf_type_to_hash_mech(mech: CK_EC_KDF_TYPE) -> KResult<CK_MECHANISM_TYPE> {
+    match mech {
+        CKD_SHA1_KDF => Ok(CKM_SHA_1),
+        CKD_SHA224_KDF => Ok(CKM_SHA224),
+        CKD_SHA256_KDF => Ok(CKM_SHA256),
+        CKD_SHA384_KDF => Ok(CKM_SHA384),
+        CKD_SHA512_KDF => Ok(CKM_SHA512),
+        CKD_SHA3_224_KDF => Ok(CKM_SHA3_224),
+        CKD_SHA3_256_KDF => Ok(CKM_SHA3_256),
+        CKD_SHA3_384_KDF => Ok(CKM_SHA3_384),
+        CKD_SHA3_512_KDF => Ok(CKM_SHA3_512),
+        _ => return err_rv!(CKR_MECHANISM_PARAM_INVALID),
+    }
+}
+
+#[derive(Debug)]
+struct ECDHOperation<'a> {
+    mech: CK_MECHANISM_TYPE,
+    kdf: CK_EC_KDF_TYPE,
+    hash: Option<CK_MECHANISM_TYPE>,
+    public: &'a [u8],
+    shared: &'a [u8],
+    finalized: bool,
+}
+
+impl ECDHOperation<'_> {
+    fn new_mechanism() -> Box<dyn Mechanism> {
+        Box::new(EccMechanism {
+            info: CK_MECHANISM_INFO {
+                ulMinKeySize: MIN_EC_SIZE_BITS as CK_ULONG,
+                ulMaxKeySize: MAX_EC_SIZE_BITS as CK_ULONG,
+                flags: CKF_DERIVE,
+            },
+        })
+    }
+
+    fn register_mechanisms(mechs: &mut Mechanisms) {
+        for ckm in &[CKM_ECDH1_DERIVE, CKM_ECDH1_COFACTOR_DERIVE] {
+            mechs.add_mechanism(*ckm, Self::new_mechanism());
+        }
+    }
+
+    fn derive_new<'a>(
+        mechanism: CK_MECHANISM_TYPE,
+        params: *const CK_ECDH1_DERIVE_PARAMS,
+    ) -> KResult<ECDHOperation<'a>> {
+        let p = unsafe { *params };
+        if p.kdf == CKD_NULL {
+            if p.pSharedData != std::ptr::null_mut() || p.ulSharedDataLen != 0 {
+                return err_rv!(CKR_MECHANISM_PARAM_INVALID);
+            }
+        }
+        if p.pPublicData == std::ptr::null_mut() || p.ulPublicDataLen == 0 {
+            return err_rv!(CKR_MECHANISM_PARAM_INVALID);
+        }
+
+        let shared = if p.pSharedData != std::ptr::null_mut()
+            && p.ulSharedDataLen != 0
+        {
+            unsafe {
+                std::slice::from_raw_parts(
+                    p.pSharedData,
+                    p.ulSharedDataLen as usize,
+                )
+            }
+        } else {
+            &[]
+        };
+
+        Ok(ECDHOperation {
+            finalized: false,
+            mech: mechanism,
+            kdf: p.kdf,
+            hash: if p.kdf == CKD_NULL {
+                None
+            } else {
+                Some(kdf_type_to_hash_mech(p.kdf)?)
+            },
+            shared: shared,
+            public: unsafe {
```

Comment:
given you are using to_vec() later, why not immediately use bytes_to_vec! here?
It is going to be safer, as it immediately alloacetes and copies the buffer, no risk the buffer disappears before it is used.

Reply:
Done.